src/Microsoft.AspNet.Mvc.Core/DefaultControllerFactory.cs

```diff
@@ -51,5 +76,90 @@ public void ReleaseController(object controller)
+
+            for (var i = 0; i < propertiesToActivate.Length; i++)
+            {
+                var activateInfo = propertiesToActivate[i];
+                activateInfo.Activate(controller, context);
+            }
+        }
+
+        /// <summary>
+        /// Returns a lookup of property types to delegates used to activate properties.
+        /// </summary>
+        /// <returns>The lookup.</returns>
+        protected virtual IDictionary<Type, Func<ActionContext, object>> CreateValueAccessorLookup()
+        {
+            var dictionary = new Dictionary<Type, Func<ActionContext, object>>
+            {
+                { typeof(ActionContext), (context) => context },
+                { typeof(HttpContext), (context) => context.HttpContext },
+                { typeof(HttpRequest), (context) => context.HttpContext.Request },
+                { typeof(HttpResponse), (context) => context.HttpContext.Response },
+                {
+                    typeof(ViewDataDictionary),
+                    (context) =>
+                    {
+                        var serviceProvider = context.HttpContext.RequestServices;
+                        return new ViewDataDictionary(
+                            serviceProvider.GetRequiredService<IModelMetadataProvider>(),
+                            context.ModelState);
+                    }
+                },
+                {
+                    typeof(ActionBindingContext),
+                    (context) =>
+                    {
+                        var serviceProvider = context.HttpContext.RequestServices;
+                        var accessor = serviceProvider.GetRequiredService<IScopedInstance<ActionBindingContext>>();
+                        return accessor.Value;
+                    }
+                }
+            };
+
+            return dictionary;
+        }
+
+        private PropertyActivator<ActionContext> CreateActivateInfo(
+            PropertyInfo property)
+        {
+            Func<ActionContext, object> valueAccessor;
+            if (!_valueAccessorLookup.TryGetValue(property.PropertyType, out valueAccessor))
+            {
+                var message = Resources.FormatControllerFactory_PropertyCannotBeActivated(
+                                                    property.Name,
+                                                    property.DeclaringType.FullName);
+                throw new InvalidOperationException(message);
+            }
+
+            return new PropertyActivator<ActionContext>(property, valueAccessor);
+        }
+
+        private PropertyActivator<ActionContext> CreateFromServicesInfo(
+            PropertyInfo property)
+        {
+            Func<ActionContext, object> valueAccessor = actionContext =>
+                    actionContext.HttpContext.RequestServices.GetService(property.PropertyType);
+
+            return new PropertyActivator<ActionContext>(property, valueAccessor);
```

Comment:
we should just cache these, it's a small non-risky change, and if model binding gets delayed we got nothing to worry about. Note that this used to be cached up until now, so we shouldn't regress perf for such a small change